File: implementations/typescript/src/delivery-journal.js
export class DeliveryJournal {
  constructor(options = {}) {
    this._streamId = options.streamId ?? "stream_01";
    this._events = [];
    this._sequence = 0;
  }

  nextSequence() {
    return ++this._sequence;
  }

  append(event) {
    const seq = this.nextSequence();
    const record = {
      ...event,
      _journal_sequence: seq,
      _journal_cursor: `${this._streamId}:${seq}`,
      _journal_appendedAt: new Date().toISOString()
    };
    this._events.push(record);
    return seq;
  }

  replay(cursor) {
    if (!cursor) return [...this._events];
    const parts = cursor.split(":");
    const sinceSeq = parseInt(parts[1] ?? "0", 10);
    return this._events.filter((e) => e._journal_sequence > sinceSeq);
  }

  replaySinceSequence(seq) {
    return this._events.filter((e) => e._journal_sequence > seq);
  }

  purge(cursor) {
    const parts = cursor.split(":");
    const beforeSeq = parseInt(parts[1] ?? "0", 10);
    let removed = 0;
    while (this._events.length > 0 && this._events[0]._journal_sequence <= beforeSeq) {
      this._events.shift();
      removed++;
    }
    return removed;
  }

  getStats() {
    return {
      totalEvents: this._events.length,
      oldestSequence: this._events.length > 0 ? this._events[0]._journal_sequence : null,
      newestSequence: this._events.length > 0 ? this._events[this._events.length - 1]._journal_sequence : null
    };
  }
}

File: implementations/java/src/main/java/com/axisrobo/harmovela/recovery/DeliveryJournal.java
package com.axisrobo.harmovela.recovery;

import java.time.Instant;
import java.util.*;

public class DeliveryJournal {
    private final String streamId;
    private final List<Map<String, Object>> events = new ArrayList<>();
    private int sequence;

    public DeliveryJournal() { this("stream_01"); }

    public DeliveryJournal(String streamId) { this.streamId = streamId; }

    public int nextSequence() { return ++sequence; }

    public int append(Map<String, Object> event) {
        var seq = nextSequence();
        var record = new LinkedHashMap<>(event);
        record.put("_journal_sequence", seq);
        record.put("_journal_cursor", streamId + ":" + seq);
        record.put("_journal_appendedAt", Instant.now().toString());
        events.add(record);
        return seq;
    }

    public List<Map<String, Object>> replay(String cursor) {
        if (cursor == null) return List.copyOf(events);
        var parts = cursor.split(":");
        var sinceSeq = parts.length > 1 ? Integer.parseInt(parts[1]) : 0;
        return events.stream().filter(e -> (int) e.get("_journal_sequence") > sinceSeq).toList();
    }

    public List<Map<String, Object>> replaySinceSequence(int seq) {
        return events.stream().filter(e -> (int) e.get("_journal_sequence") > seq).toList();
    }

    public int purge(String cursor) {
        var parts = cursor.split(":");
        var beforeSeq = parts.length > 1 ? Integer.parseInt(parts[1]) : 0;
        int removed = 0;
        while (!events.isEmpty() && (int) events.get(0).get("_journal_sequence") <= beforeSeq) {
            events.remove(0);
            removed++;
        }
        return removed;
    }

    public Map<String, Object> getStats() {
        var stats = new LinkedHashMap<String, Object>();
        stats.put("totalEvents", events.size());
        stats.put("oldestSequence", events.isEmpty() ? null : events.get(0).get("_journal_sequence"));
        stats.put("newestSequence", events.isEmpty() ? null : events.get(events.size() - 1).get("_journal_sequence"));
        return stats;
    }
}
